### ISSUE TITLE
Assertion Error in twisted.internet.abstract

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11530]
|<img alt="srichter's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| srichter reported|
|-|-|
|Trac ID|trac#986|
|Type|defect|
|Created|2005-04-17 17:42:39Z|
|Branch|https://github.com/twisted/twisted/tree/assert-986|

<details><summary>Searchable metadata</summary>

```
trac-id__986 986
type__defect defect
reporter__srichter srichter
priority__high high
milestone__ 
branch__branches_assert_986 branches/assert-986
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1113759759000000 1113759759000000
changetime__1368160749000000 1368160749000000
version__ 
owner__glyph glyph
cc__jknight cc__srichter
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="srichter's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|srichter commented|
|-|-|

```
#!html
<pre>
This is another exception I get when running web2 atop  
Zope 3: 
 
Traceback (most recent call last):  
  File "z3.py", line 59, in run  
    main(argv[1:])  
  File "/opt/zope/Zope3/Zope3-Twisted/src/zope/app/server/main.py", line 75,  
in main  
    reactor.run()  
  File "/opt/zope/Zope3/Zope3-Twisted/src/twisted/internet/posixbase.py", line  
205, in run  
    self.mainLoop()  
  File "/opt/zope/Zope3/Zope3-Twisted/src/twisted/internet/posixbase.py", line  
213, in mainLoop  
    self.runUntilCurrent()  
--- &lt;exception caught here> ---  
  File "/opt/zope/Zope3/Zope3-Twisted/src/twisted/internet/base.py", line 532,  
in runUntilCurrent  
    call.func(*call.args, **call.kw)  
  File "/opt/zope/Zope3/Zope3-Twisted/src/twisted/web2/http.py", line 974, in  
_startNextRequest  
    self.resumeProducing()  
  File "/opt/zope/Zope3/Zope3-Twisted/src/twisted/protocols/basic.py", line  
289, in resumeProducing  
    self.transport.resumeProducing()  
  File "/opt/zope/Zope3/Zope3-Twisted/src/twisted/internet/abstract.py", line  
293, in resumeProducing  
    assert self.connected and not self.disconnecting  
exceptions.AssertionError:
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Resolved by r13604.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set status to `reopened`|
|-|-|

The referenced commit had no tests, so it's no surprise that this can still happen...

I still see this error on occasion.  The changes in r13604 are to something in web2, but the assertion error is in `twisted.internet.abstract`, which web2 only interact(ed|s) with via public APIs.

You should never be able to get an assertion error out of a public API.

I'll work on a unit test to reproduce this.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

Any progress on this?  If you don't have a unit test yet, maybe just a traceback would be good information to add?  Perhaps it will inspire someone else.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

(In [38343]) Branching to 'assert-986'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

[Build results.](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/assert-986)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

And with the addition of the NEWS file, I think this is ready for review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph removed owner<br>@glyph set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @glyph|
|-|-|

Thanks.

  1. I think `set` might be a more appropriate data structure for the two new attributes of `_FakeFDSetReactor`.
  1. Please add documentation for the two new attributes of `_FakeFDSetReactor`.
  1. `IReactorFDSet` also includes the methods `getReaders` and `getWriters`.  Perhaps implementing these and then using them in the test methods would be nice.
  1. Should we do something more about uses of `resumeProducing` in this invalid state?  Presumably some application code somewhere has a bug in it that will now pass silently.  Will that make life harder for someone debugging their application?  I'm not sure what to suggest doing here though.
  1. _ought_ crept into your test method docstrings.  This is a synonym for _should_, I think (as long as we're not talking about vocabulary for RFCs).
  1. Use `assertEqual`, not `assertEquals`, in new code.
  1. You may not be surprised to learn about [twisted/internet/iocpreactor/abstract.py?rev=33856#L386 this code](twisted/internet/iocpreactor/abstract.py?rev=33856#L386 this code).  Maybe the new tests would be better if they were part of a `ReactorBuilder`-style test case - perhaps (the not extremely well named, somewhat incorrectly documented) `TCPConnectionTestsBuilder` case.  At least file a ticket for the IOCP/testing issue if you don't feel like fixing it now.
  1. I feel like `self.transport.resumeProducing()` (in the news file) is an awkward way to identify the location of the bug being fixed here.  Though I don't have a good suggestion for what to replace it with, either.

Please address the above to your satisfaction and then merge.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph commented|
|-|-|

(In [38371]) Address review point 1: make attributes of `_FakeFDSetReactor` into `set`s.

Refs [#11530](https://github.com/twisted/twisted/issues/11530)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph commented|
|-|-|

(In [38372]) Address review point 3; make `_FakeFDSetReactor` implement the whole interface
(and, as a bonus, verify it).  Use public methods instead of internal
implementation details.

Refs [#11530](https://github.com/twisted/twisted/issues/11530)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_15"></a>|@glyph commented|
|-|-|

(In [38373]) Address review point 2.  Add tons of docstrings (somewhat redundant, but that's
better than missing!)

Refs [#11530](https://github.com/twisted/twisted/issues/11530)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph commented|
|-|-|

(In [38374]) Add a test for the positive path so the test docstrings make more sense, and
address review point 5 by removing 'ought' from those docstrings.

Refs [#11530](https://github.com/twisted/twisted/issues/11530)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph commented|
|-|-|

(In [38375]) Address review point 6.  `assertEquals` -> `assertEqual`.

Refs [#11530](https://github.com/twisted/twisted/issues/11530).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph commented|
|-|-|

(In [38376]) > 7. You may not be surprised to learn about this code.

Surprised?  No.  Disappointed?  Eternally.

Re-factor the tests, trying to address this review point.  Tests fail like this,
unfortunately, because other (irrelevant) readers are in the tests.

This unfortunately obviates the need for `_FakeFDSetReactor`, but I think I'll
leave it in the merge anyway.

Refs [#11530](https://github.com/twisted/twisted/issues/11530)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph commented|
|-|-|

(In [38377]) Still trying to address point 7.  Abstract out assertions about reading; make
harness more reliable by dropping conections rather than introducing a duplicate
reactor.stop().

Refs [#11530](https://github.com/twisted/twisted/issues/11530)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_20"></a>|@glyph commented|
|-|-|

(In [38378]) Add readability assertions for IOCP.

Did I mention that I was disappointed?  We apparently don't have an interface
for `addActiveHandle`/`removeActiveHandle`, nor an analagous `getActiveHandles`
method.

Refs [#11530](https://github.com/twisted/twisted/issues/11530)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_21"></a>|@glyph set owner to @exarkun|
|-|-|

All right, that was a bit too much churn for me to be comfortable just landing.

Please give it a quick once-over and I can land if you think it's reasonable.

To make your job easier, [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/assert-986).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_22"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_12):

Finally, regarding this:

> Should we do something more about uses of `resumeProducing` in this invalid state?  Presumably some application code somewhere has a bug in it that will now pass silently.  Will that make life harder for someone debugging their application?  I'm not sure what to suggest doing here though.

I think _maybe_ a warning is appropriate, but I'm not entirely sure that would actually be helpful.  Do we care about the same issue with `loseConnection`?  It has the same issue.  `reactor.stop`, which "helpfully" raises an exception, is an unbearable nuisance, and I wish that just passed silently too.

In any case this should be a separate ticket; the current behavior is undocumented and arguably unintentional.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun set owner to @glyph<br>@exarkun set status to `new`|
|-|-|

Thanks.

> Add tons of docstrings (somewhat redundant, but that's better than missing!)

Regarding this, take a look at what some people did for [#6459](https://github.com/twisted/twisted/issues/6459) and think about which way you prefer.

> We apparently don't have an interface for addActiveHandle/removeActiveHandle, nor an analagous getActiveHandles method.

Indeed.  There's also `_getWriters` in twisted/internet/connectionmixins.py.  Maybe we should have a ticket for this and these special cases should link to it.  

Just one other thing - I think we (as in you and I) might have talked about the idea that led you to make `assertReading` and the other new functions here top-level instead of methods.  But this idiom probably merits some explanation in the code.  More generally, maybe the next area we need to improve is how we well explain the big picture behind the tiny fragments of differentness that we sometimes start to introduce.  How can we get some other developer to continue this pattern in the direction we want?

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/assert-986) are ... disappointing ... but the problems don't look directly attributable to this change.  Please merge after addressing the minor points above.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_25"></a>|@glyph commented|
|-|-|

(In [38382]) Taking my lead from [#6459](https://github.com/twisted/twisted/issues/6459), don't document things that already have documentation.

This is an interesting (and good) feature of pydoctor, but it's a little bit
unfortunate while reading the code itself.  I wonder if maybe in the future we
ought to implement (or suggest to the Zope Interface maintainers that they
implement) an `@implementation` decorator that specifically calls out a method
as an implementation of an interface method that should not be documented on its
own.

Refs [#6459](https://github.com/twisted/twisted/issues/6459)

Refs [#11530](https://github.com/twisted/twisted/issues/11530)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_26"></a>|@glyph commented|
|-|-|

(In [38383]) An attempt in communicating some design intent, as well as updating the name and
role of one of the parameters.

Refs [#11530](https://github.com/twisted/twisted/issues/11530)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_27"></a>|@glyph set status to `closed`|
|-|-|

(In [38384]) Merge assert-986

Remove an `assert` both "abstract" (therefore: TCP) implementations of
`resumeProducing`.

Author: glyph

Reviewer: exarkun

Fixes: [#11530](https://github.com/twisted/twisted/issues/11530)

self.transport.resumeProducing() will no longer raise an AssertionError if called after self.transport.loseConnection()
